CIP-0071/README.md

```diff
@@ -0,0 +1,246 @@
+
+This specification follows the same process, but using tokens:
+1. A holder of a project validates their NFT by sending it to self
+2. The holder signs a Plutus minting policy to create a "ballot" NFT linked to their unique NFT
+3. The holder marks their desired vote selections off-chain
+4. The holder signs a tx that sends the "ballot" NFT to a "ballot box" (smart contract) with their "vote" (datum)
+5. Authorized vote counting wallets process UTxOs and their datums in the "ballot box" smart contract after polls close
+6. Authorized vote counters report the results in a human-readable off-chain format to holders
+
+Because of the efficient UTxO model Cardano employs, steps #1, #2, and #4 can occur in a single transaction.
+
+### The Vote Casting Process
+
+#### "Ballot" -> Plutus Minting Policy
+
+Every holder that participates in the vote will have their project NFT in a wallet that can be spent from (either hardware or software, typically accessed via [CIP-30](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030)).  To create a ballot, the voting authority will create a Plutus minting policy with a specific combination of:
+
+```ts
+type BallotMintingPolicy = {
+  referencePolicyId: MintingPolicyHash,             // Reference policy ID of the original NFT project
+  pollsClose: Time,                                 // Polls close (as a Unix timestamp in milliseconds)
+  assetNameMapping: func(ByteArray) -> ByteArray    // Some function (potentially identity) to map reference NFT name 1-for-1 to ballot NFT name
+};
+```
+
+This Plutus minting policy will perform the following checks:
+1. Polls are still open during the Tx validFrom/validTo interval
+2. The reference NFT assets were spent (typically, sent-to-self) for each ballot being minted
+3. The minted assets are sent directly to the ballot box smart contract in the minting transaction (see [the potential attack below](#Creation-of-Ballot-Without-Casting-a-Vote))
+
+For the voter, each vote they wish to cast will require creating a separate "ballot" NFT.  In the process, their reference NFT never leaves the original wallet.  Sample [Helios language](https://github.com/Hyperion-BT/Helios) pseudocode (functions elided for space) is as follows:
+
+```ts
+func main(redeemer: Redeemer, ctx: ScriptContext) -> Bool {
+  tx: Tx = ctx.tx;
+  minted_policy: MintingPolicyHash = ctx.get_current_minting_policy_hash();
+  redeemer.switch {
+    Mint => {
+      polls_are_still_open(tx.time_range)
+        && assets_were_spent(tx.minted, minted_policy, tx.outputs)
+        && assets_locked_in_script(tx, tx.minted)
+    },
+    // Burn code elided for space...
+  }
+}
+```
+
+#### "Vote" -> eUTxO Datum
+
+To cast the vote, the user sends the ballot NFT just created to a "ballot box".  Note that for reasons specified in [the "attacks" section below](#Creation-of-Ballot-Without-Casting-a-Vote) this needs to occur during the same transaction that the ballot was minted in.
```

Comment:
What enforces that? The minting policy?

Reply:
Yes that is the `assets_locked_in_script` function above (will rename to `assets_locked_in_ballot_box`).